### PR TITLE
Accessibility: Adding Keyboard Shorcuts to navigate the editor regions

### LINKED ISSUE #467: Add keyboard shortcuts à la Slack to jump through the main UI sections
Reference: looking at the screenshot posted on Slack on https://wordpress.slack.com/archives/C02QB2JS7/p1492621873449513

The toolbar:

screen shot 2017-04-19 at 19 07 00

The order and logical grouping of controls/links/info is fundamental for accessibility, and also usability. There are many things to consider here, will try to highlight some general issues.

**The Publish button:**
The Editor experience should be a `type type type` experience. For accessibility, add to that some `tab tab tab` experience. 🙂  Keyboard users and screen reader users will expect the Publish button to be _after_ the editing area. That's the convention for any form. Placing it before the editing area would repeat the same mistake done in the Customizer with the "Save" button. Forcing users to tab backwards or explore the content to find the Publish button would be extremely confusing and annoying.
For keyboard and screen reader users, navigating content is a linearized experience that follows the order of elements in the markup. So the Publish button should come after the editor in the markup.
Worth also noting that the WCAG state the visual order should match the tab order.
I'd say this is also an usability issue, since the logical workflow when editing a post is:
Enter the title > Enter the content > Publish

**The other controls/info etc:**
The position of the other controls should be carefully considered too. Probably some of them can be placed before the editor. However, the most important thing here is they should be logically grouped.
I know this is just a first implementation and many things are probably temporary, but worth considering that currently these controls are a bit mixed up and there's no a clear, logical grouping.
Trying to list their different purposes:
```
1: action related to the editing tool:      switch the Editor's mode
2: info related to the editing state:       Saved (not sure if it's meant for other purposes too)
3: action related to the editing flow:      undo/redo
4: action related to the editing tool/flow: a "Plus" thing, I guess to show additional editing controls
5: link related to the editing state:       preview the post
6: action related to the editing tool/flow: control to expand the Post Settings
7: action related to the editing flow:      Publish button
```

I may be wrong on some of them 🙂  but my goal here is to highlight they have a different nature and purpose, so they should be grouped accordingly. Off the top of my head, a possible grouping could be:
1. actions that do something to the content (undo, redo, etc.)
2. actions that do something to the editing tool (showing additional controls, switching mode, etc.)
3. things that show the post state (Saved, Preview, etc.)
4. the Publish button is a special case (see above)

I'd say the group 1. should be placed as close as possible to the content, in a place that's very easy to reach when tabbing. These are controls that get used often while editing content so reaching them should require a minimal effort.

Groups 2. and 3. are probably going to be used not so often as the first group. I'm curious about the plus "+" control though, since it's a bit unclear what it will reveal when pressed. Same for the "Post Settings".

Some of the issues about "what's around the Editor" were discussed also on https://core.trac.wordpress.org/ticket/29838 where one of things considered was to remove anything that comes _before_ the editor and place it elsewhere. This is especially true when it comes to mobile devices. I'd recommend everyone to follow the links in the ticket and watch the videos / read the posts by Mark Jaquith and Ryan Boren.

Additional considerations:
- where's the "Add Media" button?
- where's a space dedicated to plugins that add their custom buttons? (usually, they add them after the "Add Media" one)

### PR DESCRIPTION
**Pull request by youknowriad**
closes #467 

This PR adds keyboard shortcuts to navigate the aria regions ala slack.

For now, I came up with random shortcuts (`ctrl+r` and `ctrl+shift+r`), let me know which one you think I should use.

### CODE REVIEW

**Comment by jasmussen** [comment]
Totally love this! Very very nice work. Finding a good standardized shortcut for this is going to be important, especially given the backtick discussio in last weeks meeting. 

I like the focus style for the region — is it kosher to do something like this? https://codepen.io/joen/pen/Nvjror — that is, a big flashing focus style that disappears?

I don't know if it's intentional — I assume it is, but the shortcut doesn't work if you have focus inside a text block. 

Nice work!

**Comment by afercia** [comment]
Love to see this in action 😄 Thanks!
(and started considering to propose this for the WordPress admin too)

Shortcut: yee we should find a good one. I'd love to use `G` as **G**utenberg ;) But I guess we should use a combination with a non a-z 0-9 key. @joedolson help please.

@jasmussen no objections to experiment the flashing focus limited to this case. I'd like to see this merged soon and ready for testing. We can always review later if necessary.

> I don't know if it's intentional — I assume it is, but the shortcut doesn't work if you have focus inside a text block.

Noticed that and I understand that's maybe a bit tricky to implement. However, if the idea of the blocks "Navigation mode" and "Edit mode" discussed during last SLack chat will be implemented, then makes sense to me having these shortcuts working just in "Navigation mode".

Instead, when editing and focus is inside a block, the process would be:
press Escape > exit the edited block and goes back to "Navigation mode" > shortcuts work again
Thoughts?

Aside: we should document all this in a future Gutenberg help guide.

**Comment by mtias** [comment]
> Aside: we should document all this in a future Gutenberg help guide.

Good call, we should start writing these sooner rather than later. They could even be extracted to be part of the "new user experience" cc @karmatosed

File: editor/layout/index.js [hunk 1]
```diff
@@ -28,29 +29,77 @@ import {
...
-			<DocumentTitle />
-			<NoticeList onRemove={ props.removeNotice } notices={ notices } />
-			<UnsavedChangesWarning />
-			<AutosaveMonitor />
-			<Header />
-			<div className="editor-layout__content">
-				<div className="editor-layout__editor">
-					{ mode === 'text' && <TextEditor /> }
-					{ mode === 'visual' && <VisualEditor /> }
+class Layout extends Component {
```
Comment (mtias) [inline editor/layout/index.js]: Looking at the previous simple component, wonder if we should convert the region-focus into a HoC that could be used in different editor layouts.

**Comment by jasmussen** [comment]
> Shortcut: yee we should find a good one. I'd love to use G as Gutenberg ;) But I guess we should use a combination with a non a-z 0-9 key. 

Question: can we use _both_ the standardized backtick shortcut for keyboards that support this, _and_ a better one for us europeans? (perhaps Ctrl+$, as $ is right above tab)? The idea being that if backtick is a standard for this, good to embrace it for the majority of the world, with enhancements for the rest of us?

> no objections to experiment the flashing focus limited to this case. I'd like to see this merged soon and ready for testing. We can always review later if necessary.

Cool, Riad can we try the animated focus style from that codepen I linked? Let me know if you'd like me to implement it. It uses $blue-medium-500.

> Good call, we should start writing these sooner rather than later. They could even be extracted to be part of the "new user experience" cc @karmatosed

Would a good start be to implement the keyboard shortcuts cheatsheet (invoked by for example Shift + ?) as ticketed in #2980 be a good start? 

I don't know how far we can go with the componentization thing, but it would be really nice if, when a shortcut is used, the tooltip and cheat sheet inclusion create themselves.

File: editor/layout/index.js [hunk 2]
```diff
@@ -28,29 +29,77 @@ import {
...
+						{ mode === 'visual' && <VisualEditor /> }
+					</div>
+					<MetaBoxes location="normal" />
 				</div>
-				<MetaBoxes location="normal" />
+				{ isSidebarOpened && <Sidebar /> }
+				<Popover.Slot />
+				<KeyboardShortcuts shortcuts={ {
+					'ctrl+r': this.focusNextRegion,
+					'ctrl+shift+r': this.focusPreviousRegion,
```
Comment (afercia) [inline editor/layout/index.js]: On Windows, Ctrl+r refreshes the page, see extended comment on the PR.

**Comment by afercia** [comment]
Testing on Windows, we've missed that Ctrl+r refreshes the page. The default action should be prevented but...

we should find a better shortcut anyways. Finding a good one that is safe to use across different browsers, platforms, and keyboard layouts is basically impossible. 

> perhaps Ctrl+$, as $ is right above tab

Not on my Italian keyboard, maybe that's the Danish Mac keyboard 😉. I guess $ is placed differently also in other keyboard layouts.

If we decide these shortcuts should be available only in "navigation mode" (i.e. not when editing a block or when inside any other form field) then I'd consider to use single key shortcuts.

To my understanding, modifiers like `Control`, `Option`, `Shift`, are necessary when inside a form field. Instead, outside of form fields, single keys should work because browsers will ignore single key presses. Thoughts?

Many big apps already use single keys shortcuts, for example Gmail uses `j` and `k` to navigate through previous/next messages, and those are pretty common shortcuts. Testing on Gmail:
- open a message in your inbox
- press `j` or `k` to navigate through messages
- click the "Compose" button to open the new message modal
- in the new message form fields, `j` or `k` do nothing of course
- press Escape to close the modal
- `j` and `k` work again

Maybe replicating this interaction would make sense also in Gutenberg? As in, when editing a block, press "Escape" exits edit mode and goes back to navigation mode, then the shortcuts work again. Note: it already works this way.

WordPress already uses single key shortcuts for the Comments screen: https://codex.wordpress.org/Keyboard_Shortcuts#Keyboard_Shortcuts_for_Comments

We don't have to worry about screen readers because they already have their mechanism to jump through landmark regions. Windows screen readers use "browse mode" and "forms mode". When in "browse mode" the Gutenberg shortcuts won't do anything because screen readers intercepts keypresses, so keypresses are not passed to the browser. This is not a problem because, for example, NVDA and JAWS already have their shortcuts:

NVDA list of single key shortcuts:
https://www.nvaccess.org/files/nvda/documentation/userGuide.html#toc41
> d: landmark

JAWS list of single key shortcuts:
https://www.freedomscientific.com/Training/Surfs-Up/Quick_Keys.htm
> Region	R

I'd propose to try `j` as in "Jump" because it's easy to remember.

Note: the new meta boxes expanding panel is not included in an ARIA landmark region. Please consider, also for future developments, that all content in a web page should be inside a landmark region. Yes, it's inside the WordPress "main" landmark but the point is if this content section is important enough to be a navigable region. Should we make it a landmark adding `role="region" aria-label="some label"`? Should it be focusable and included in the shortcuts loop?
Want me to open a separate issue?

**Comment by jasmussen** [comment]
I love direct shortcuts like j/k, I'd wager they'd be really popular if we were able to add a bunch of those. 

However I've seen the Ctrl+` shortcut used as a standard to navigate between _views in an application_ in many places. MacOS has a ⌘+` shortcut for switching between _windows_ in an application (i.e. if you have two Safari windows open, ⌘+` switches between those two, _because for whatever dumb reason the Mac designers thought Alt + Tab should flip between only applications, not windows_).

And so in my googling for european alternatives to the backtick, I realized that the backtick on US keyboards sits right above the tab button. That's where I have the $ key. And I figured ... hmmm... could they have? And sure enough, Ctrl+$ moves focus from region to region in Slack. 

Andrea can you test whether Ctrl + _the key right above your tab button, below Esc_ switches between regions in Slack? If yes, then that's how they did it — they chose the key by geography. And if it works for you — should we consider doing the same?

File: editor/layout/index.js (hunk 1 above)
Comment (youknowriad) [reply editor/layout/index.js]: Do you think this should be a generic component in the `components` module? It looks like something that could be useful even outside an editor context.

Comment (youknowriad) [reply editor/layout/index.js]: done in daa9cf3

**Comment by mtias** [comment]
In my Spanish keyboard it works and is represented as:

image

**Comment by youknowriad** [comment]
Wondering how do we match a keyboard position regardless of the layout. Is this the same keycode even if it's a different key? In my french keyboard it's `@` and it works in slack.

If I try in this website http://keycode.info/ it gives me a keycode of 192. Are you seeing the same keycode?

**Comment by jasmussen** [comment]
I would _guess_ that we have to put together an array of keycodes for this character, based on whatever info we can glean.

**Comment by afercia** [comment]
> Andrea can you test whether Ctrl + the key right above your tab button, below Esc switches between regions in Slack?

I'm afraid Slack implemented an alternate shortcuts only for the languages Slack is translated in. Spanish, French, and German. On an Italian keyboard layout, the key right above your tab button, below Esc is "Back slash" and does nothing. The Slack help keeps showing me Ctrl + backtick.

Switching to French does show me: Ctrl  + @ 
Switching to German does show me: Ctrl  + ^
Wasn't able to get the Spanish shortcut though.

> I realized that the backtick on US keyboards sits right above the tab button

Not on all US keyboards, apparently. On my mac US keyboard is right above Ctrl:

screen shot 2017-10-23 at 16 47 47

Id recommend to keep it simple 😉

**Comment by mtias** [comment]
@youknowriad I do get 192 for `º`.

**Comment by youknowriad** [comment]
Should we go with keycode 192 whatever that is on each keyboard layout :)?

I think we won't be able to use the `KeyboardShortcuts` component and mousetrap to match this but not a big issue.

**Comment by jasmussen** [comment]
If we are able to go with multiple shortcuts for the same, it seems we might be able to make the most people happy by picking a shortcut that...

1. sits at a convenient location on the keyboard
2. is a standard

While 2 isn't universally true, until we find a single keyboard shortcut that's _actually_ a standard, the closest is what we have here with ⌘+` that Slack uses. It seems worth trying this.

We could then supplement it, or replace it, with a simpler shortcut that doesn't change between keyboards, if we can find a good idea for what that should be.

**Comment by youknowriad** [comment]
Interestingly I've used ⌘+` as a shortcut using the `mousetrap` lib in it's working properly with `⌘+@` in my keyboard. Can y'all try now?

**Comment by youknowriad** [comment]
And it's a flashing focus now :)

**Comment by jasmussen** [comment]
Ctrl + $ (the key above my tab) isn't working for me, and Ctrl + @ doesn't work for me as @ requires an Alt for me.

**Comment by youknowriad** [comment]
🤔 Can you try this "String.fromCharCode(192)" on your console and tell me what it gives you?

**Comment by jasmussen** [comment]
screen shot 2017-10-26 at 10 33 44

So that's `À`

**Comment by youknowriad** [comment]
@jasmussen Oh you're on a mac right? did you try `command + $` instead?

**Comment by jasmussen** [comment]
Oh interesting. I didn't try that, and I have that mapped to something else. But Macs also have Ctrl, and slack on Mac also uses Ctrl for this. I think CMD + backtick on a Mac is supposed to flip between windows of the same app

**Comment by youknowriad** [comment]
Replaced command by ctrl, does it work now?

**Comment by jasmussen** [comment]
It works for me now, and it's completely awesome. Given that the flashing also happens when you just _click_ to set tab focus somewhere, I think it's good that it's so fast and disappears after a while. The focus in the main content area is cropped at the top — not a huge issue but if it's an easy fix then would be nice to get in:

focus

**Comment by youknowriad** [comment]
@jasmussen I suspect the cropping to be fixed once we land this commit a0fdfc749958f642689d9885c2762dc0e1ef2a83 from #2998

**Comment by jasmussen** [comment]
Then 👍 👍  for the feature from me. I'll let others chime in on the shortcuts and implementation, but I dig this implementation.

File: editor/assets/stylesheets/_animations.scss [hunk 3]
```diff
@@ -38,3 +38,23 @@
 	50% { opacity: 1; }
 	100% { opacity: .5; }
 }
+
+
+@mixin focus_flash {
+  animation: focus_flash .4s;
```
Comment (mtias) [inline editor/assets/stylesheets/_animations.scss]: Are spaces mixed in here?

File: editor/layout/index.js [hunk 4]
```diff
@@ -60,4 +60,4 @@ export default connect(
 		notices: getNotices( state ),
 	} ),
 	{ removeNotice }
-)( Layout );
+)( navigateRegions( Layout ) );
```
Comment (mtias) [inline editor/layout/index.js]: Looks much better, thanks.

File: editor/assets/stylesheets/_animations.scss (hunk 3 above)
Comment (youknowriad) [reply editor/assets/stylesheets/_animations.scss]: Good catch 👍

**Comment by youknowriad** [comment]
Rebased this, it should be ready to go

**Comment by mtias** [comment]
The cropping on the main area doesn't to be fixed after the rebase? Testing this, I feel that the fact it disappears so quickly doesn't communicate what this does very well—almost feels like a glitch.

**Comment by youknowriad** [comment]
Oh!! I think this broke somehow with the rebase. looking

**Comment by youknowriad** [comment]
@mtias Still investigating but you can try with "shift", it works better.

**Comment by youknowriad** [comment]
Ok per @mtias suggestion, I dropped the focus animation for now

**Comment by mtias** [comment]
Feels much better to me like this, though now it exposes the crop issue more clearly when you scroll:

image

**Comment by youknowriad** [comment]
Good catch, it should be fixed now

**Comment by mtias** [comment]
Looks great to me.

**Comment by afercia** [comment]
I've tested a bit on a Windows laptop with the Italian keyboard. There's no physical key for the backtick. On Italian keyboards with a numeric pad, the only way to produce a backtick is pressing Alt+96 on the numeric pad. Many laptops don't even have a numeric pad...

Pressing the key below Escape (in my case it's a back slash) does nothing.
Only when I switch the keyboard layout to English the key below Escape works as backtick and the regions navigation works.

Will open a new issue.

**Comment by afercia** [comment]
Just had confirmation from @rianrietveld the backtick doesn't work on the Dutch keyboard layout (and as said previously, probably on many other layouts). This makes https://github.com/WordPress/gutenberg/issues/3218 really necessary.

**Pull request merged**